--- cryptobox/a/PKEY.ec	2026-09-10 17:48:14
+++ cryptobox/b/PKEY.ec	2026-09-10 17:51:22
@@ -363,6 +363,7 @@
     + by move=> pk _; rewrite dmap1E dkp_max.
   move=> [] pk sk /= _.
   by rewrite mem_fdom domE filterE=> ->.
+  by move=> &hr _; smt(ge0_pr_guess fcard_ge0).
 by hoare; auto=> />.
 qed.
 
@@ -785,6 +786,7 @@
     + by move=> pk _; rewrite dmap1E dkp_max.
   move=> [] pk sk /= _.
   by rewrite mem_fdom domE filterE /is_honest; case: skm.[pk]=> [|[]] /#.
+  by move=> &hr _; smt(ge0_pr_guess fcard_ge0).
 by hoare; auto=> /#.
 qed.
 
