diff --git a/proof/SHA3OSecurity.ec b/proof/SHA3OSecurity.ec
index e8babf0..3921f92 100644
--- a/proof/SHA3OSecurity.ec
+++ b/proof/SHA3OSecurity.ec
@@ -1791,7 +1791,9 @@ have->:Pr[SecondPreimage(A, OSponge, PSome(Perm)).main(mess) @ &m : res] =
 + by byequiv=> //=; proc; inline*; auto; sim.
 byphoare(: arg = mess ==>_)=>//=; proc; sp.
 call(: arg = mess /\ mess = Dist_of_P2Adv.m ==> res); auto.
-bypr=> {&m} &m [#]->; rewrite eq_sym=> Heq.
+bypr=> {&m} &m; split.
++ by smt(sigma_ge0 StdOrder.IntOrder.expr_gt0 StdOrder.RealOrder.expr_gt0 Ring.IntID.expr2).
+move=> [#]->; rewrite eq_sym=> Heq.
 rewrite (rw_real &m mess Heq).
 have := SHA3OIndiff (Dist_of_P2Adv(A)) &m _.
 + move=> F P Hp Hpi Hf; proc; inline*; sp; auto; call Hf; auto; call Hf; auto. 
diff --git a/proof/SecureORO.eca b/proof/SecureORO.eca
index e244982..8810616 100644
--- a/proof/SecureORO.eca
+++ b/proof/SecureORO.eca
@@ -125,7 +125,7 @@ section Preimage.
     bypr=> /> {&m} &m.
     have H: forall &m h, 
          Pr[FEL(A, RF(RO)).main(h) @ &m : rng RO.m h] <= bound%r * mu1 sampleto witness; last first.
-    + exact(H &m).
+    + by split; [smt(bound_ge0 ge0_mu1 mu_bounded) | exact(H &m)].
     move=> {&m h} &m h.
     fel 1 Bounder.bounder (fun _, mu1 sampleto witness) bound (rng RO.m h) 
       [Bounder(RF(RO)).get: (card (fdom RO.m) <= Bounder.bounder < bound)]
@@ -138,7 +138,7 @@ section Preimage.
     - inline*; auto=> />.
       rewrite mem_rng_empty /= fdom0 fcards0 /=; smt(bound_ge0). 
     - proc.
-      sp; if; auto; sp; inline*; sp; wp=> /=.
+      sp; if; auto; 2:(by move=> &hr; smt(ge0_mu1 mu_bounded)); sp; inline*; sp; wp=> /=.
       case: (x \in RO.m); wp => //=.
       + by hoare; auto; smt(mu_bounded). 
       rnd (pred1 h); auto=> /> &h c H0 H1 H2 H3 H4 H5.
@@ -280,7 +280,8 @@ section SecondPreimage.
     auto=> /> sample.
     by rewrite mem_set mem_empty/= fdom_set fdom0 fset0U fcard1; smt(bound_ge0).
   + call(: arg = mess1 ==> rng (rem RO.m mess1) (oget RO.m.[mess1])); auto.
-    bypr=> {&m} &m h; rewrite h.
+    bypr=> {&m} &m; split; 1: smt(bound_ge0 ge0_mu1 mu_bounded).
+    move=> h; rewrite h.
     fel 2 Bounder.bounder (fun _, mu1 sampleto witness) bound 
         (mess1 \in RO.m /\ rng (rem RO.m mess1) (oget RO.m.[mess1]))
         [Bounder(RF(RO)).get: (card (fdom RO.m) - 1 <= Bounder.bounder < bound)]
@@ -444,21 +445,25 @@ section Collision.
       by move=> in_dom1; smt().
     by move=> />; rewrite fdom0 fcards0; smt(bound_ge0).
   call(: true ==> collision RO.m); auto; bypr=> /> {&m} &m.
-  fel 1 Bounder.bounder (fun i, i%r * mu1 sampleto witness) bound 
+  split; 1: smt(bound_ge0 ge0_mu1 mu_bounded).
+  fel 1 Bounder.bounder (fun i, (max 0 i)%r * mu1 sampleto witness) bound 
         (collision RO.m)
         [Bounder(RF(RO)).get: (card (fdom RO.m) <= Bounder.bounder < bound)]
         (card (fdom RO.m) <= Bounder.bounder <= bound)=> //. 
-  + rewrite -StdBigop.Bigreal.BRA.mulr_suml RField.mulrAC.
+  + rewrite -StdBigop.Bigreal.BRA.mulr_suml (StdBigop.Bigreal.BRA.eq_big_seq _ (%r)); 1: smt(List.Range.mem_range).
+    rewrite RField.mulrAC.
     rewrite StdOrder.RealOrder.ler_wpmul2r; 1: smt(mu_bounded).
     by rewrite StdBigop.Bigreal.sumidE //; smt(bound_ge0).
   + inline*; auto=> />.
     rewrite fdom0 fcards0; split; 2: smt(bound_ge0). 
     rewrite /collision negb_exists /= => a; rewrite negb_exists /= => b.
     by rewrite mem_empty.
-  + bypr=> /> {&m} &m; pose c := Bounder.bounder{m}; move=> H0c Hcbound Hcoll Hmc _.
+  + bypr=> /> {&m} &m; split; 1: smt(ge0_mu1 mu_bounded).
+    pose c := Bounder.bounder{m}; move=> H0c Hcbound Hcoll Hmc _.
+    have ->: max 0 c = c by smt().
     byphoare(: !collision RO.m /\ card (fdom RO.m) <= c ==> _)=>//=.
     proc; inline*; sp; if; last first.
-    - by hoare; auto=> _ _; smt(le_fromint ge0_mu).
+    - by hoare; auto; smt(le_fromint ge0_mu).
     case: (x \in RO.m).
     - by hoare; auto; smt(le_fromint mu_bounded).
     rcondt 5; 1: auto; sp; wp=> /=.
diff --git a/proof/SecureRO.eca b/proof/SecureRO.eca
index f6b4e5e..8cd076d 100644
--- a/proof/SecureRO.eca
+++ b/proof/SecureRO.eca
@@ -126,6 +126,7 @@ section Preimage.
       by rewrite fdom_set fcardU fcard1; smt(fcard_ge0).
     by auto=> />; rewrite fdom0 fcards0; smt(bound_gt0).
   + call(: true ==> rng RO.m h)=> //; bypr=> /> {&m} &m.
+    split; 1: smt(bound_gt0 ge0_mu1 mu_bounded).
     fel 1 Bounder.bounder (fun _, mu1 sampleto h) bound (rng RO.m h) 
       [Bounder(RO).get: (card (fdom RO.m) <= Bounder.bounder < bound)]
       (card (fdom RO.m) <= Bounder.bounder <= bound)
@@ -137,7 +138,7 @@ section Preimage.
     - inline*; auto=> />. 
       by rewrite mem_rng_empty /= fdom0 fcards0 /=; smt(bound_gt0). 
     - proc.
-      sp; if; auto; sp; inline*; sp; wp=> /=.
+      sp; if; auto; 2:(by move=> &hr; smt(ge0_mu1 mu_bounded)); sp; inline*; sp; wp=> /=.
       if; last by hoare; auto; progress; smt(mu_bounded).
       case: (x \in RO.m); wp => //=.
       + by hoare; auto; smt(mu_bounded). 
@@ -300,7 +301,8 @@ section SecondPreimage.
     auto=> /> sample.
     by rewrite mem_set mem_empty/= fdom_set fdom0 fset0U fcard1; smt(bound_gt0).
   + call(: arg = mess1 ==> rng (rem RO.m mess1) (oget RO.m.[mess1])); auto.
-    bypr=> {&m} &m h; rewrite h.
+    bypr=> {&m} &m; split; 1: smt(bound_gt0 ge0_mu1 mu_bounded).
+    move=> h; rewrite h.
     fel 2 Bounder.bounder (fun _, mu1 sampleto witness) bound 
         (mess1 \in RO.m /\ rng (rem RO.m mess1) (oget RO.m.[mess1]))
         [Bounder(RO).get: (card (fdom RO.m) - 1 <= Bounder.bounder < bound)]
@@ -477,18 +479,22 @@ section Collision.
       by move=> in_dom1; smt().
     by move=> />; rewrite fdom0 fcards0; smt(bound_gt0).
   call(: true ==> collision RO.m); auto; bypr=> /> {&m} &m.
-  fel 1 Bounder.bounder (fun i, i%r * mu1 sampleto witness) bound 
+  split; 1: smt(bound_gt0 ge0_mu1 mu_bounded).
+  fel 1 Bounder.bounder (fun i, (max 0 i)%r * mu1 sampleto witness) bound 
         (collision RO.m)
         [Bounder(RO).get: (card (fdom RO.m) <= Bounder.bounder < bound)]
         (card (fdom RO.m) <= Bounder.bounder <= bound)=> //. 
-  + rewrite -StdBigop.Bigreal.BRA.mulr_suml RField.mulrAC.
+  + rewrite -StdBigop.Bigreal.BRA.mulr_suml (StdBigop.Bigreal.BRA.eq_big_seq _ (%r)); 1: smt(List.Range.mem_range).
+    rewrite RField.mulrAC.
     rewrite StdOrder.RealOrder.ler_wpmul2r; 1: smt(mu_bounded).
     by rewrite StdBigop.Bigreal.sumidE //; smt(bound_gt0).
   + inline*; auto=> />.
     rewrite fdom0 fcards0; split; 2: smt(bound_gt0). 
     rewrite /collision negb_exists /= => a; rewrite negb_exists /= => b.
     by rewrite mem_empty.
-  + bypr=> /> {&m} &m; pose c := Bounder.bounder{m}; move=> H0c Hcbound Hcoll Hmc _.
+  + bypr=> /> {&m} &m; split; 1: smt(ge0_mu1 mu_bounded).
+    pose c := Bounder.bounder{m}; move=> H0c Hcbound Hcoll Hmc _.
+    have ->: max 0 c = c by smt().
     byphoare(: !collision RO.m /\ card (fdom RO.m) <= c ==> _)=>//=.
     proc; inline*; sp; if; last first.
     - by hoare; auto; smt(le_fromint ge0_mu).
diff --git a/proof/smart_counter/Gcol.eca b/proof/smart_counter/Gcol.eca
index 918c7cd..70bb0a7 100644
--- a/proof/smart_counter/Gcol.eca
+++ b/proof/smart_counter/Gcol.eca
@@ -206,7 +206,7 @@ section PROOF.
     + rewrite /felsum Bigreal.sumr_const count_predT size_range.
       apply ler_wpmul2r;1:by apply eps_ge0.
       by rewrite le_fromint;smt ml=0 w=max_ge0. 
-    + proc;sp;if;2:by hoare=>//??;apply eps_ge0.
+    + proc;sp;if;2:(by hoare=>//; smt(eps_ge0)).
       wp.
       rnd (mem (image fst (frng FRO.m)));skip;progress;2:smt ml=0. 
       have->:=(Mu_mem.mu_mem (image fst (frng FRO.m{hr})) cdistr (1%r/(2^c)%r) _).
